### ISSUE TITLE
bogus error "redeclaration with a different type" for redefined auto-typed inline variable

### ISSUE THREAD

**Issue by zygoloid** [llvm/llvm-project#214560]
Testcase:
```c++
#pragma clang module build std
module std {
    module a {}
    module b {}
}
#pragma clang module contents
#pragma clang module begin std.a
inline constexpr auto x = [] {};
#pragma clang module end
#pragma clang module begin std.b
inline constexpr auto x = [] {};
#pragma clang module end
#pragma clang module endbuild
```

is rejected with this error:

```console
std.map:10:23: error: redeclaration of 'x' with a different type: 'const (lambda at std.map:10:27)' vs 'const (lambda at std.map:7:27)'
   10 | inline constexpr auto x = [] {};
      |                       ^
std.map:7:23: note: previous definition is here
    7 | inline constexpr auto x = [] {};
      |                       ^
```

I believe this is incorrect: the ODR says the two lambdas introduce the same type, because they are within the definition of the same inline variable. Our definition merging logic is presumably not handling this case properly.

In practice, this can happen in some build configurations within libc++, here:
```console
libcxx/include/__compare/synth_three_way.h:28:45: error: redeclaration of '__synth_three_way' with a different type: 'const std::__u::(lambda at libcxx/include/__compare/synth_three_way.h:28:65)' vs 'const std::__u::(lambda at libcxx/include/__compare/synth_three_way.h:28:65)'
```

Presumably the same thing would happen in a C++20 build:
```c++
export module A;
extern "C++" inline auto a = [] {};
```
```c++
module B;
import A;
extern "C++" inline auto a = [] {};
```
or indirectly via:
```c++
module;
#include <compare>
export module A;
```
```c++
import A;
#include <compare>
```
(if you don't have a modular libc++).

**Comment by zygoloid** [comment]
Weirdly, we don't get this error in the C++20 modules versions I posted; only in the module map variant. Possibly the bug was fixed but only for C++20 modules?

**Comment by zygoloid** [comment]
Looks like it's not C++20 modules versus Clang modules that's making the difference, it's whether the lambda was parsed locally versus imported from an AST file. We handle the imported-lambda case here:

https://github.com/llvm/llvm-project/blob/5e8ac69abe478b1702000878d47fddb6234af725/clang/lib/Serialization/ASTReader.cpp#L9873

... which is wrong; we need the same handling regardless of whether the lambda comes from an AST file. Presumably `LambdaDefinitionsForMerging` needs to live on the `ASTContext` rather than the `ASTReader`.
